### PR TITLE
fix(mget): log when compact object type changes and add a fallback mechanism to recheck

### PR DESCRIPTION
**Pull request by kostasrim**

### CODE REVIEW

File: src/core/compact_object.cc [hunk 1]
```diff
@@ -1012,7 +1012,8 @@ void CompactObj::GetString(char* dest) const {
 
   // no encoding.
   if (taglen_ == ROBJ_TAG) {
-    CHECK_EQ(OBJ_STRING, u_.r_obj.type());
+    // TODO bring this check back once we find the root cause that this fails
```
Comment (romange) [inline src/core/compact_object.cc:1015]: please revert this change

File: src/server/string_family.cc [hunk 2]
```diff
@@ -525,13 +525,32 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
   response.storage_list = SinkReplyBuilder::AllocMGetStorage(total_size);
   char* next = response.storage_list->data;
 
+  auto log_error = [](auto it) {
+    LOG(ERROR) << "Invalid type " << it->second.ObjType();
+    LOG(ERROR) << "Bucket id: " << it.bucket_id();
+    LOG(ERROR) << "Segment " << it.segment_id();
+    LOG(ERROR) << "Slot id " << it.slot_id();
```
Comment (chakaz) [inline src/server/string_family.cc:532]: maybe log expected key and key found, I suspect they won't be identical

File: src/server/string_family.cc [hunk 3]
```diff
@@ -525,13 +525,32 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+      // OpResult<PrimeConstIterator> it_res =
+      //  db_slice.FindAndFetchReadOnly(t->GetDbContext(), args[i], OBJ_STRING);
+      //// if we failed continue
+      // if (!it_res && it->second.ObjType() != OBJ_STRING) {
+      //   continue;
+      // }
+      //// update iterators if got the correct type
+      // iters[i] = *it_res;
+      // it = iters[i];
+    }
```
[the remark this answers is not in the document]
Comment (chakaz) [reply src/server/string_family.cc:553]: I think that a retry could spiral out, we already are in an UB mode (not C++ but logical)
I'd just continue

File: src/server/string_family.cc [hunk 4]
```diff
@@ -525,13 +525,32 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+  };
+
   for (size_t i = 0; i < args.size(); ++i) {
     PrimeConstIterator it = iters[i];
     if (it.is_done())
       continue;
 
     auto& resp = response.resp_arr[i].emplace();
-
+    if (it->second.ObjType() != OBJ_STRING) {
```
Comment (romange) [inline src/server/string_family.cc:541]: 1. I would like to print all the valid iterators in  iters, not just the one that is inconsistent
2. I actually want to test two conditions: objtype like you do here and also it->first vs `args[i]`. i.e. if any of them do not hold, you trigger the fallback.
3. The fallback should also call again `db_slice.FindReadOnly` and find the correct iterator and continue with the normal flow of checking again  for ObjType and it.done and proceeding with CopyValueToBuffer again.

Comment (romange) [reply src/server/string_family.cc:541]: The code should not be beautiful, but should be correct. It's a temporary code.

File: src/server/string_family.cc [hunk 5]
```diff
@@ -525,11 +525,41 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+
   for (size_t i = 0; i < args.size(); ++i) {
     PrimeConstIterator it = iters[i];
     if (it.is_done())
       continue;
 
+    std::string key_found;
+    it->first.GetString(&key_found);
+
+    log_error(it, args[i], key_found);
```
Comment (kostasrim) [inline src/server/string_family.cc:545]: Unconditional log

Comment (romange) [reply src/server/string_family.cc:545]: I was not clear. I would like to log everything (all the iterators) only if there is a mismatch and we enter line 548.

File: src/server/string_family.cc [hunk 6]
```diff
@@ -525,11 +525,41 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+    if (it->second.ObjType() != OBJ_STRING || key_found != args[i]) {
+      // Do the checks again
+      OpResult<PrimeConstIterator> it_res =
+          db_slice.FindAndFetchReadOnly(t->GetDbContext(), args[i], OBJ_STRING);
+      if (it_res->is_done())
```
Comment (romange) [inline src/server/string_family.cc:550]: also LOG here that we can not find `args[i]` anymore

File: src/server/string_family.cc [hunk 7]
```diff
@@ -525,11 +525,41 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+        continue;
+      if (!it_res && it->second.ObjType() != OBJ_STRING) {
```
Comment (romange) [inline src/server/string_family.cc:552]: you should swap the order between this condition and is_done before

File: src/server/string_family.cc [hunk 8]
```diff
@@ -525,11 +525,41 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+        continue;
```
Comment (romange) [inline src/server/string_family.cc:553]: also log here the reason

File: src/server/string_family.cc [hunk 9]
```diff
@@ -525,11 +525,41 @@ SinkReplyBuilder::MGetResponse OpMGet(bool fetch_mcflag, bool fetch_mcver, const
...
+      }
+      // we continue here because it's hard to readjust the size of response.storage_list in line
+      // 525 An alternative solution would be to allocate more (2x) and remove the check
+      if (it->second.Size() != (*it_res)->second.Size()) {
```
Comment (romange) [inline src/server/string_family.cc:557]: please remove this check, it's wrong.
